### PR TITLE
[pipeline] OPT model pipeline

### PR DESCRIPTION
**Pull request by CjhHa1**
## 📌 Checklist before creating the PR

- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [ ] I have created an issue for this PR for traceability
- [x] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`
#4219 

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.
This pr adds opt model pipeline and policy

## 💥 Checklist before requesting a review

- [x] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [x] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [x] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

**Review by Fridge003 — APPROVED** [review]

File: colossalai/shardformer/policies/opt.py [hunk 1]
```diff
@@ -114,10 +183,29 @@ def module_policy(self):
...
+        held_layers = super().get_held_layers()
+        if self.pipeline_stage_manager.is_last_stage():
+            held_layers.append(self.model.lm_head)
+        return held_layers
+
+    def get_shared_params(self) -> List[Dict[int, Tensor]]:
+        opt_model = self.model
+        if self.pipeline_stage_manager:
+            num_stages = self.pipeline_stage_manager.num_stages
+            if id(opt_model.model.decoder.embed_tokens.weight) == id(opt_model.lm_head.weight):
```
Comment (ver217) [inline colossalai/shardformer/policies/opt.py:204]: Check whether these params are in the different stages here.

File: colossalai/shardformer/policies/opt.py [hunk 2]
```diff
@@ -135,8 +223,659 @@ class OPTForSequenceClassificationPolicy(OPTPolicy):
...
+        mask_cond = torch.arange(mask.size(-1), device=device)
+        mask.masked_fill_(mask_cond < (mask_cond + 1).view(mask.size(-1), 1), 0)
+        mask = mask.to(dtype)
+
+        if past_key_values_length > 0:
+            mask = torch.cat([torch.zeros(tgt_len, past_key_values_length, dtype=dtype, device=device), mask], dim=-1)
+        return mask[None, None, :, :].expand(bsz, 1, tgt_len, tgt_len + past_key_values_length)
+
+    @staticmethod
+    def _expand_mask(mask: torch.Tensor, dtype: torch.dtype, tgt_len: Optional[int] = None):
```
Comment (ver217) [inline colossalai/shardformer/policies/opt.py:323]: Do you modify this method?

Comment (CjhHa1) [reply colossalai/shardformer/policies/opt.py:323]: the method _expand_mask is not modified

Comment (CjhHa1) [reply colossalai/shardformer/policies/opt.py:305]: done

File: colossalai/shardformer/policies/opt.py (hunk 1 above)
Comment (CjhHa1) [reply colossalai/shardformer/policies/opt.py:204]: done

**Review by Fridge003 — APPROVED** [review]

**Pull request merged**